### PR TITLE
Karma: waitfor implementation improvements part one

### PR DESCRIPTION
**Pull request by BrookeGraham**
## Context

This is to clean up some karma tests that were not using `waitFor` correctly. 

## Relevant Technical Choices

In places where we were using `waitFor` in addition to `getBy`, converted to using  await `findBy` since ```findBy methods are a combination of getBy queries and waitFor. They accept the waitFor options as the last argument```.  When that approach doesn't suffice, we throw an error if what we are waiting for isn't ready.

## Todo
Broke up the pr into two in an attempt to make reviewing easier. Additional pr #10429 😉

Found a test that is failing due to poor implementation. The test was grabbing the wrong element, which was leading to the appearance of it being flakey. Created a bug ticket #10428 

- [x] getting a false positive for canvas keys, fix dialog issue.  

## Testing Instructions

- [x] This is a non-user-facing change and requires no QA

This PR can be tested by following these steps:

1.

## Checklist

- [x] This PR addresses an existing issue and I have linked this PR to it in ZenHub
- [x] I have tested this code to the best of my abilities
- [x] I have verified accessibility to the best of my abilities ([docs](https://github.com/googleforcreators/web-stories-wp/blob/main/docs/accessibility-testing.md))
- [x] I have verified i18n and l10n (translation, right-to-left layout) to the best of my abilities
- [x] This code is covered by automated tests (unit, integration, and/or e2e) to verify it works as intended ([docs](https://github.com/googleforcreators/web-stories-wp/tree/main/docs#testing))
- [x] I have added documentation where necessary
- [x] I have added a matching `Type: XYZ` label to the PR

---

Partially addresses #10194

### CODE REVIEW

File: packages/dashboard/src/app/views/myStories/karma/myStories.karma.js [hunk 1]
```diff
@@ -127,7 +127,7 @@ describe('CUJ: Creator can view their stories in grid view', () => {
 
       await fixture.events.click(exploreTemplatesMenuItem);
       const templatesGridEl = await waitFor(() =>
-        fixture.screen.queryByText('Viewing all templates')
+        fixture.screen.findByText('Viewing all templates')
```
Comment (swissspidy) [inline packages/dashboard/src/app/views/myStories/karma/myStories.karma.js]: From the [Testing Library docs](https://testing-library.com/docs/dom-testing-library/api-async#findby-queries):

> `findBy` methods are a combination of `getBy` queries and `waitFor`. They accept the `waitFor` options as the last argument

File: packages/story-editor/src/components/canvas/karma/keys.karma.js [hunk 2]
```diff
@@ -112,9 +112,17 @@ describe('Canvas keys integration', () => {
...
+    await waitFor(
+      () => {
+        const dialog = fixture.container.querySelector('[role="dialog"]');
+        if (!dialog) {
+          throw new Error('dialog not ready');
+        }
+        expect(dialog).toBeTruthy();
+      }
+      // { timeout: 4000 }
+    );
```
Comment (swissspidy) [inline packages/story-editor/src/components/canvas/karma/keys.karma.js]: Could perhaps replace this with `findByRole('dialog')` instead?

File: packages/dashboard/src/app/views/myStories/karma/myStories.karma.js (hunk 1 above)
Comment (BrookeGraham) [reply packages/dashboard/src/app/views/myStories/karma/myStories.karma.js]: Oh thats right! Thanks!

File: packages/story-editor/src/components/library/panes/text/karma/textSets.cuj.karma.js [hunk 3]
```diff
@@ -83,15 +102,24 @@ describe('CUJ: Text Sets (Text and Shape Combinations): Using Text Sets', () =>
...
     // Text sets contain at least 2 elements.
     expect((await getSelection()).length).toBeGreaterThan(1);
   });
 
   it('should allow user to drag and drop text set onto page', async () => {
     await waitFor(
-      () => expect(fixture.editor.library.text.textSets.length).toBeTruthy(),
+      () => {
+        if (fixture.editor.library.text.textSets.length === 0) {
+          throw new Error('test sets not ready');
```
Comment (BrookeGraham) [inline packages/story-editor/src/components/library/panes/text/karma/textSets.cuj.karma.js]: typo

File: packages/story-editor/src/components/panels/design/link/karma/link.karma.js [hunk 4]
```diff
@@ -150,17 +162,27 @@ describe('Link Panel', () => {
...
-      });
+      await waitFor(
+        () => {
+          const tooltip = fixture.screen.findByText(linkDescription);
+          if (!tooltip) {
+            throw new Error('tooltip not ready');
+          }
+          expect(tooltip).toHaveTextContent(linkDescription);
+        },
+        { timeout: 5000 }
```
Comment (BrookeGraham) [inline packages/story-editor/src/components/panels/design/link/karma/link.karma.js]: Why did I add this 🤔 Note to self: double check it is still needed.

File: packages/story-editor/src/components/canvas/karma/keys.karma.js [hunk 5]
```diff
@@ -112,9 +107,9 @@ describe('Canvas keys integration', () => {
     await fixture.events.click(colorButton);
     expect(await getSelection()).toEqual([element1.id]);
 
-    await waitFor(() => fixture.container.querySelector('[role="dialog"]'), {
-      timeout: 4000,
-    });
+    const dialog = fixture.screen.findByRole('dialog');
```
Comment (BrookeGraham) [inline packages/story-editor/src/components/canvas/karma/keys.karma.js]: add `await`

File: packages/story-editor/src/components/canvas/karma/quickActions.karma.js [hunk 6]
```diff
@@ -641,15 +652,17 @@ describe('Quick Actions integration', () => {
...
-      await waitFor(
-        () =>
-          fixture.screen.getByRole('button', { name: /^Undo$/, hidden: true }),
-        { timeout: 4000 }
-      );
-
+      await fixture.screen.findByRole('button', {
+        name: /^Undo$/,
+        hidden: true,
+        timeout: 400,
```
Comment (BrookeGraham) [inline packages/story-editor/src/components/canvas/karma/quickActions.karma.js]: add zero

File: packages/story-editor/src/components/canvas/karma/quickActions.karma.js [hunk 7]
```diff
@@ -823,15 +836,18 @@ describe('Quick Actions integration', () => {
...
-      await waitFor(
-        () =>
-          fixture.screen.getByRole('button', { name: /^Undo$/, hidden: true }),
-        { timeout: 4000 }
-      );
 
+      await fixture.screen.findByRole('button', {
+        name: /^Undo$/,
+        hidden: true,
+        timeout: 400,
```
Comment (BrookeGraham) [inline packages/story-editor/src/components/canvas/karma/quickActions.karma.js]: add zero, or if possible remove extra timeout setting to use default

File: packages/story-editor/src/components/canvas/karma/quickActions.karma.js [hunk 8]
```diff
@@ -1041,15 +1060,17 @@ describe('Quick Actions integration', () => {
...
-      await waitFor(
-        () =>
-          fixture.screen.getByRole('button', { name: /^Undo$/, hidden: true }),
-        { timeout: 4000 }
-      );
-
+      await fixture.screen.findByRole('button', {
+        name: /^Undo$/,
+        hidden: true,
+        timeout: 400,
```
Comment (BrookeGraham) [inline packages/story-editor/src/components/canvas/karma/quickActions.karma.js]: here too

File: packages/story-editor/src/components/canvas/karma/quickActions.karma.js [hunk 9]
```diff
@@ -1205,15 +1229,17 @@ describe('Quick Actions integration', () => {
...
```
Comment (BrookeGraham) [inline packages/story-editor/src/components/canvas/karma/quickActions.karma.js]: again

File: packages/story-editor/src/components/canvas/karma/keys.karma.js [hunk 10]
```diff
@@ -112,9 +107,9 @@ describe('Canvas keys integration', () => {
     await fixture.events.click(colorButton);
     expect(await getSelection()).toEqual([element1.id]);
 
-    await waitFor(() => fixture.container.querySelector('[role="dialog"]'), {
-      timeout: 4000,
-    });
+    const dialog = await fixture.screen.findByRole('dialog');
+    expect(dialog).toBeTruthy();
+
```
Comment (BrookeGraham) [inline packages/story-editor/src/components/canvas/karma/keys.karma.js]: There is more than one dialog.

File: packages/story-editor/src/components/library/panes/text/karma/textPane.karma.js [hunk 11]
```diff
@@ -122,13 +137,22 @@ describe('CUJ: Creator can Add and Write Text: Consecutive text presets', () =>
...
+        });
         nodeIndex++;
         storyContext = await fixture.renderHook(() => useStory());
-        const element = storyContext.state.selectedElements[0];
+        let element = null;
         await waitFor(() => {
+          element = storyContext.state.selectedElements[0];
+          if (!element) {
+            throw new Error('story not ready');
+          }
```
Comment (BrittanyIRL) [inline packages/story-editor/src/components/library/panes/text/karma/textPane.karma.js:155]: this is super interesting to have to do! I get that element's not quite set yet every time, but it definitely drives home why we've all scattered so many `wait waitFor WAIT PLEASE 😉 ` everywhere.

**Review by BrittanyIRL — APPROVED** [review]
Thank you for cleaning all of this up!! Tests are working as expected and the updated patterns make sense.

File: packages/story-editor/src/components/library/panes/media/media3p/karma/mediaFetching.karma.js [hunk 12]
```diff
@@ -267,29 +267,19 @@ describe('Media3pPane fetching', () => {
...
-
-        if (!mediaElements || mediaElements.length === 0) {
-          throw new Error('mediaElements did not load');
-        }
-      },
-      { timeout: 5000 }
-    );
+    const mediaElements = await within(
+      fixture.editor.library.media3p.unsplashSection
+    ).findAllByTestId(/^mediaElement/, { timeout: 5000 });
```
Comment (samwhale) [inline packages/story-editor/src/components/library/panes/media/media3p/karma/mediaFetching.karma.js:272]: this is cool, I never knew that you could do this!

File: packages/story-editor/src/components/colorPicker/karma/colorPicker.karma.js [hunk 13]
```diff
@@ -52,9 +52,12 @@ describe('ColorPicker', () => {
...
 
-        await waitFor(() =>
-          expect(bgPanel.backgroundColor.picker).toBeDefined()
-        );
+        await waitFor(() => {
+          if (!bgPanel.backgroundColor.picker) {
+            throw new Error('picker not ready');
+          }
+          expect(bgPanel.backgroundColor.picker).toBeDefined();
+        });
```
Comment (samwhale) [inline packages/story-editor/src/components/colorPicker/karma/colorPicker.karma.js:60]: 💭  What are your thoughts on changing the queries in the containers themselves to be `findBy` but use a max timeout of `5000ms`?

That way when people use these, they won't have to worry about making sure they are using `waitFor` correctly?

image

**Review by samwhale — APPROVED** [review]
Thanks for doing all of this. Super tedious stuff. Also, cool to see that `findBy` is a way to use `waitFor` 😄

File: packages/story-editor/src/components/canvas/karma/keys.karma.js [hunk 14]
```diff
@@ -111,10 +106,10 @@ describe('Canvas keys integration', () => {
     );
     await fixture.events.click(colorButton);
     expect(await getSelection()).toEqual([element1.id]);
+    // there are two dialogs, [0] checklist companion and [1]color picker
+    const dialogs = await fixture.screen.findAllByRole('dialog');
+    expect(dialogs.length).toBe(2);
 
-    await waitFor(() => fixture.container.querySelector('[role="dialog"]'), {
-      timeout: 4000,
-    });
```
Comment (BrookeGraham) [inline packages/story-editor/src/components/canvas/karma/keys.karma.js:117]: Interestingly I found that the checklist companion dialog was always (at least when I was debugging here) present. So when we would check for role=‘dialog’ we were often getting the checklist dialog and not whatever one we were actually trying to target. ie. color picker.

File: packages/story-editor/src/components/colorPicker/karma/colorPicker.karma.js (hunk 13 above)
Comment (BrookeGraham) [reply packages/story-editor/src/components/colorPicker/karma/colorPicker.karma.js:60]: I like the idea.

**Pull request merged**